larpandoracontent/LArObjects/LArAdaBoostDecisionTree.cc

```diff
@@ -0,0 +1,299 @@
+/**
+ *  @file   larpandoracontent/LArObjects/LArAdaBoostDecisionTree.cc
+ *
+ *  @brief  Implementation of the lar adaptive boost decision tree class.
+ *
+ *  $Log: $
+ */
+
+#include "Helpers/XmlHelper.h"
+
+#include "larpandoracontent/LArObjects/LArAdaBoostDecisionTree.h"
+
+namespace lar_content
+{
+
+AdaBoostDecisionTree::AdaBoostDecisionTree() :
+    m_isInitialized(false)
+{
+    std::cout << "AdaBoostDecisionTree::AdaBoostDecisionTree" << std::endl;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+pandora::StatusCode AdaBoostDecisionTree::Initialize(const std::string &parameterLocation, const std::string &bdtName)
+{
+    if (m_isInitialized)
+    {
+        std::cout << "AdaBoostDecisionTree: bdt was already initialized" << std::endl;
+        return pandora::STATUS_CODE_FAILURE;
+    }
+
+    // ATTN: Add consistency check for number of features, bearing in min not all features in a bdt may be used
+    this->ReadXmlFile(parameterLocation, bdtName);
+ 
+    m_isInitialized = true;
+    return pandora::STATUS_CODE_SUCCESS;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+bool AdaBoostDecisionTree::Classify(const DoubleVector &features) const
+{
+    return (m_pStrongClassifier->Predict(features) > 0.0 ? true : false);
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+double AdaBoostDecisionTree::CalculateClassificationScore(const DoubleVector &features) const
+{
+    return m_pStrongClassifier->Predict(features);
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+double AdaBoostDecisionTree::CalculateProbability(const DoubleVector &features) const
+{
+    return (m_pStrongClassifier->Predict(features) + 1.0) * 0.5;
```

Comment:
Where does this come from?

Reply:
Normalising the BDT score by the weights, which I believe is standard procedure in things like TMVA, produced BDT scores between -1 and 1.  This linear mapping changes that to give a score between 0 and 1, which we can interpret as a probability.  I don't know if a more appropriate mapping is possible, but this is what's done in the pandora SVM use case also.

Reply:
Probably best to add an `// ATTN` comment to explain this.

Reply:
Done.